model-server/src/main/kotlin/org/modelix/model/server/handlers/ui/DiffView.kt

```diff
@@ -0,0 +1,680 @@
+                                createLinkToContentExplorer(repositoryId, newVersionHash, change.newParent)
+                            }
+                            td {
+                                +"${change.newRole}"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun FlowContent.buildConceptChangesSection(
+        diff: VersionDiff,
+        repositoryId: String,
+        newVersionHash: String,
+    ) {
+        div {
+            id = "conceptChanges"
+            if (diff.conceptChanges.isEmpty()) {
+                p { +"No concept changes." }
+                return@div
+            }
+            h2 { +"Concept Changes" }
+            table {
+                createTableHead(listOf("Node", "Old Concept", "New Concept"))
+                tbody {
+                    for (change in diff.conceptChanges) {
+                        tr {
+                            td {
+                                createLinkToContentExplorer(repositoryId, newVersionHash, change.node)
+                            }
+                            td { +"${change.oldConcept}" }
+                            td { +"${change.newConcept}" }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun TABLE.createTableHead(headings: List<String>) {
+        thead {
+            tr {
+                for (heading in headings) {
+                    td { +heading }
+                }
+            }
+        }
+    }
+
+    private fun FlowContent.createLinkToContentExplorer(
+        repositoryId: String,
+        versionHash: String,
+        node: CPNode,
+    ) {
+        a(constructUrlToContentExplorer(repositoryId, versionHash, node)) {
+            +"${node.id}${node.getNameClarification()}"
+        }
+    }
+
+    private fun constructUrlToContentExplorer(
+        repositoryId: String,
+        versionHash: String,
+        node: CPNode? = null,
+    ): String {
+        val url = "$baseUrl/content/repositories/${repositoryId.encodeURLPath()}/versions/${versionHash.encodeURLPath()}/"
+        return if (node == null) {
+            url
+        } else {
+            url + "?expandTo=${node.id}"
+        }
+    }
+}
+
+private fun CPNode.getNameClarification(): String {
+    val name = getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name.getUID())
+        ?: getPropertyValue("name")
+
+    return name?.let { "($it)" }.orEmpty()
+}
+
+private fun CLTree.resolveOrThrow(nodeId: Long): CPNode =
+    requireNotNull(resolveElement(nodeId)) { "node not found. id = $nodeId" }
+
+private fun CLTree.tryResolve(ref: CPNodeRef?): CPNode? {
+    return try {
+        resolveElement(ref)
+    } catch (e: RuntimeException) {
+        null
+    }
+}
+
+internal fun calculateDiff(oldVersion: CLVersion, newVersion: CLVersion, sizeLimit: Int = Int.MAX_VALUE): VersionDiff {
```

Comment:
Consider breaking up `DiffView` into `DiffView` and `DiffCalculation`.
Same for the tests.

Reply:
I was debating that as well. I put them into one file since they are closely related (see [coding conventions](https://kotlinlang.org/docs/coding-conventions.html#source-file-organization)). However one could argue that the file has become slightly too large.

Reply:
Ah, they are closely related for now.
Keeping them in on file is fine.